design/events.md

```diff
@@ -0,0 +1,417 @@
+      <td>
+        <ul>
+          <li>Branch name</li>
+          <li>Previous HEAD</li>
+          <li>New HEAD</li>
+          <li><code>CommitMeta</code> object</li>
+        </ul>
+      </td>
+      <td>Operations (PUTs and DELETEs) are not included, in order to avoid huge event payloads. 
+          Full operation contents are sent separately as content-related events.
+        <p>A <code>COMMIT</code> event should trigger:</p>
+        <ul>
+          <li>0-N content events (one for each PUT/DELETE operation).</li>
+        </ul>
+      </td>
+    </tr>
+    <tr>
+      <td><code>MERGE</code></td>
+      <td>API call</td>
+      <td>
+        <ul>
+          <li>Source branch name</li>
+          <li>Source branch HEAD</li>
+          <li>Target branch name</li>
+          <li>Previous target branch HEAD</li>
+          <li>New target branch HEAD</li>
+          <li>Common ancestor hash</li>
+        </ul>
+      </td>
+      <td>Operations (PUTs and DELETEs) are not included, in order to avoid huge event payloads. 
+          Full operation contents are sent separately as content-related events.
+        <p>A <code>MERGE</code> event should trigger:</p>
+        <ul>
+          <li>0-N content events (one for each global PUT/DELETE operation, excluding intermediary operations).</li>
+        </ul>
+      </td>
+    </tr>
+    <tr>
+      <td><code>TRANSPLANT</code></td>
+      <td>API call</td>
+      <td>
+        <ul>
+          <li>Source branch name</li>
+          <li>Target branch name</li>
+          <li>Previous target branch HEAD</li>
+          <li>New target branch HEAD</li>
+        </ul>
+      </td>
+      <td>Operations (PUTs and DELETEs) are not included, in order to avoid huge event payloads. 
+          Full operation contents are sent separately as content-related events.
+        <p>A <code>TRANSPLANT</code> event should trigger:</p>
+        <ul>
+          <li>0-N content events (one for each global PUT/DELETE operation, excluding intermediary operations).</li>
+        </ul>
+      </td>
+    </tr>
+    <tr>
+      <td colspan="2" rowspan="2">Content-related events</td>
+      <td><code>CONTENT_STORED</code></td>
+      <td>Committing event</td>
+      <td>
+        <ul>
+          <li>Branch name</li>
+          <li>Commit Hash</li>
+          <li>Content type</li>
+          <li>Content key</li>
+          <li><code>Content</code> object</li>
+        </ul>
+      </td>
+      <td>Corresponds to a PUT operation from a committing event. Note that it's not possible to distinguish a creation 
+          from an update.</td>
+    </tr>
+    <tr>
+      <td><code>CONTENT_REMOVED</code></td>
+      <td>Committing event</td>
+      <td>
+        <ul>
+          <li>Branch name</li>
+          <li>Commit Hash</li>
+          <li>Content type</li>
+          <li>Content key</li>
+        </ul>
+      </td>
+      <td>Corresponds to a DELETE operation from a committing event. Note that the actual Content object that was 
```

Comment:
That's an interesting remark indeed. A rename is done in a commit containing two operations:

1. DELETE("keyOld")
2. PUT("keyNew", oldContent)

Such a rename commit would be translated into the following 3 events:

1. COMMIT
2. CONTENT_REMOVED: keyOld
3. CONTENT_STORED: keyNew, oldContent

So, it's not immediately clear that the CONTENT_REMOVED and CONTENT_STORED events form a rename.

But I'd also point out that the rename intent is not clear in the commit either: one needs to inspect the content ids to tell if it's a rename or just a deletion and a creation of two unrelated contents.

I suggest we keep the design as is for now, but if users need it, we could later on introduce a CONTENT_RENAMED event.

Wdyt? Or should we introduce it right now?

Reply:
> As long as we have the content-ID for the PUT operations, it should be fine though.

Now that I think of it... this would require a read of the commitlog, right? As we need to know if the content existed before but with a different key. Or maybe this is possible with the new storage model without additional reads? If not, I'm afraid this won't scale well. Probably not a good idea to include a RENAME event after all.

> I was initially thinking we could also publish the commit message (along with the existing commit hash)

We will. So you will see the special message mentioning the rename. That's the workaround users will need to use imho.

Reply:
> [content-ID ...] would require a read of the commitlog, right?

Is definitely there for commits. For updates, those are passed from the clients, for creates, there's a callback from the VersionStore impl. Merges and transplants... those are definitely around somewhere in the version store implementations, so should be relatively easy to expose those.